### ISSUE TITLE
Nested facet column cannot refer to nested data source for cardinality

### ISSUE THREAD

**Issue by kanitw** [vega/vega#952]
In [this spec](https://pastebin.com/Znth2HEM), we get `Undefined data set name: "child_column_layout"` although we have properly defined `child_column_layout` as a nested data source.

This is an output for the following VL spec:

```
{
  "$schema": "https://vega.github.io/schema/vega-lite/v2.json",
  "data": {"url": "data/cars.json"},
  "facet": {"column": {"field": "Origin","type": "ordinal"}},
  "spec": {
    "facet": {"column": {"field": "Cylinders","type": "ordinal"}},
    "spec": {
      "mark": "point", 
      "encoding": {
        "x": {"field": "Horsepower","type": "quantitative"},
        "y": {"field": "Acceleration","type": "quantitative"}
      }
    }
  }
}
```

**Comment by jheer** [comment]
This is not a Vega bug, though it is indeed potentially confusing. Similar to the `encode` block, the `layout` block is defined on the top-level mark items - *prior* to generating the contained scope with signals, data sets, etc. So the layout does not have access to "local" signals or data sets because it is actually defined in the outer scope. This is not just an issue of parsing either; the same layout instance is used for *all* sibling group marks. Thus, the number of columns must be the same and can not dynamically change across sibling marks. So in this case, I think you want to compute the number of distinct cylinders in an outer scope, for all the data, and then use that number.

The alternative - changing the semantics of Vega - would be a very significant undertaking at this point.

Also, a tangential side note on the implementation strategy used here: The `child_column` data set generates an aggregate that has the counts for each cylinder. You don't need to run an additional count distinct aggregate to get the number of cylinders, as that is simply the number of tuples in the `child_column` data set. So this could simplify to a signal: `length(data('child_column'))`. Of course, the issue above renders this point moot, but I thought I would share as it may inform other parts of the system.

**Comment by kanitw** [comment]
>  the same layout instance is used for all sibling group marks.

This mean we won't really truly support multi-level nesting for facet? 

Basically we can nest row and y and column and x, but not really nest row with inner row.

I wonder if this could be addressed in a later version of Vega so it does not delay our VG3.0/VL 2.0 release.

**Comment by jheer** [comment]
It depends on how we implement this. For example, I could imagine a "hack" in which each group item might have a "columns" property (set within an encode block) that could be used to override the default layout columns setting. However, the group encode block is also defined in *outside* the scope of nested signals and data sets. So, we could create dynamic column counts pretty easily, but we'd still need to determine the cardinality in the outer scope. Perhaps we could do this by adding appropriate aggregates directly to the facet output?

**Comment by kanitw** [comment]
I see.  If u want to use this hack, adding aggregate distinct calculation in facet is easy as I already have a code path to do it for calculating size signal with range step .

If this (or any other solution you prefer) isn't too hacky and not too difficult it would be great to include it in 3.0 too. Otherwise we will have to do some temporary restructure to calculate number of column in the global scope, which is quite complicated to do and may not worth it if we will have to revert it soon after we have support for nesting case

**Comment by jheer** [comment]
Added in vega/vega-view@199554cf2876d49a072b2db99f8d37323e67eb12, released in vega-view v1.2.6.
